### PR TITLE
604 add url params query

### PR DESCRIPTION
**Pull request by IsaacJenkinsTPZ**
Added the functionality to add parameters to the URL and read them in where required. Parameters saved are the view (map or list), the selected catalogue, the filters in the filter sidebar, the AOI, the map centre point and the map zoom.

I have added two implementations to this problem, the first one follows the ticket https://telespazio-uk.atlassian.net/browse/EODHP-604 and uses a "/" separated URL structure to format these parameters. The issue with this is that in order to ensure that we have the correct URL structure each time, there are cases where we have "//" in the URL for instance if there was no selected catalogue then it would be blank in the URL.

After discussion with Mark and very quickly with James, it seems that it is more common, simpler to implement and follows better practices to use query strings in the URL. These are easier to read in the code and simpler to pick out of the URL.

Both have been added in this PR, but I am have implemented the query strings method.

Let me know which one you want to keep and I can remove the one which we decide is not going to be used

(we also needed to change some of the .env parameters too so this works on dev)

### CODE REVIEW

**Review by ahaywardtvuk — COMMENTED** [review]
I'll leave @james-hinton to go through the code in detail, but I've noticed some things that are not working in dev (it might be that some are not in this repo):
* If I go to https://dev.eodatahub.org.uk/news/article-2/#test-anchor and then log in I'm returned to https://dev.eodatahub.org.uk/news/article-2/ . ie, the anchor is stripped.
* I then went to https://dev.eodatahub.org.uk/about/, signed in, and was taken to the wrong URL entirely, https://dev.eodatahub.org.uk/news/article-2/
* If I open the catalogue, go to the list view and then either reload or copy the URL into another tab then I'm taken to the map view.
* There are two parameters for the catalogue path and one is not doing anything. For example, if I go to https://dev.eodatahub.org.uk/catalogue/?catalogPath=supported-datasets to browse the Supported Datasets catalog and then click on CMIP6 I end up at https://dev.eodatahub.org.uk/catalogue/?catalogPath=supported-datasets&view=map&centre=-10018.754171394621%2C6711113.243704713&catalogue=cmip6&aoi=-0.7843354010581977%2C52.64474793150097%2C-0.07242214322090161%2C52.64474793150097%2C-0.07242214322090161%2C53.06401359958602%2C-0.7843354010581977%2C53.06401359958602%2C-0.7843354010581977%2C52.64474793150097 which has an additional `catalogue=cmip6` query parameter. This is not removed when I click 'Return to Collections'. It also doesn't designate a catalog (it's a collection) and is not enough information to know which items it should be showing when reloaded. It sounds like you need a separate catalog path which determines what is shown in the catalogs/collections tree, and collection path which determines that we're showing an items list from a certain collection.

Note that at present the catalogue path never changes after the page is loaded. This may change, but only if a user explicitly 'descends' into a sub-Catalog or chooses to go to the details page for a Collection or Item. The catalogue path determines the API endpoint the app is querying and nothing more.

We might later revisit how the URL paths are generated if you don't put the catalogue path and view type in them now. The idea would be to have URLs that mirror the API structure like this:

* https://eodatahub.org.uk/catalogue/map - map view against API endpoint https://eodatahub.org.uk/api/catalogue/stac
* (probably where 'Catalogue' will link to) https://eodatahub.org.uk/catalogue/supported-datasets/map - map view against API endpoint https://eodatahub.org.uk/api/catalogue/stac/catalogs/supported-datasets 
* https://eodatahub.org.uk/catalogue/catalogs/user-datasets/catalogs/ahayward/map - map view against my private workspace catalogue at https://eodatahub.org.uk/api/catalogue/stac/catalogs/user-datasets/catalogs/ahayward/
* (not implemented yet) https://eodatahub.org.uk/catalogue/catalogs/supported-datasets/catalogs/ceda/collections/cmip6 - dataset details page for the CMIP6 dataset

Every entry in the catalogue then has an API URL like `https://eodatahub.org.uk/api/catalogue/stac/<catalogue path>` and a UI URL like `https://eodatahub.org.uk/catalogue/<catalogue path>` , with the two things marked `<catalogue path>` being the same. Then, in the UI, if the catalogue path identifies a STAC Catalog then appending `/map` or `/list` to it gives a map or a list of its contents.

**Comment by james-hinton** [comment]
> * If I go to https://dev.eodatahub.org.uk/news/article-2/#test-anchor and then log in I'm returned to https://dev.eodatahub.org.uk/news/article-2/ . ie, the anchor is stripped.

Looks like this is fixed, or I can't recreate this bug.

> * I then went to https://dev.eodatahub.org.uk/about/, signed in, and was taken to the wrong URL entirely, https://dev.eodatahub.org.uk/news/article-2/

Looks like this is fixed, or I can't recreate this bug.

> * If I open the catalogue, go to the list view and then either reload or copy the URL into another tab then I'm taken to the map view.

This is an outstanding issue. It would appear that the `<MapViewer />` page would always be overlaying the List view regardless of what the active page is. I know @mark-small had a reason for this, but I can't remember what it was. https://github.com/EO-DataHub/eodhp-resource-catalogue-ui/blob/main/src/App.tsx#L40

> * There are two parameters for the catalogue path and one is not doing anything. For example, if I go to https://dev.eodatahub.org.uk/catalogue/?catalogPath=supported-datasets to browse the Supported Datasets catalog and then click on CMIP6 I end up at https://dev.eodatahub.org.uk/catalogue/?catalogPath=supported-datasets&view=map&centre=-10018.754171394621%2C6711113.243704713&catalogue=cmip6&aoi=-0.7843354010581977%2C52.64474793150097%2C-0.07242214322090161%2C52.64474793150097%2C-0.07242214322090161%2C53.06401359958602%2C-0.7843354010581977%2C53.06401359958602%2C-0.7843354010581977%2C52.64474793150097 which has an additional `catalogue=cmip6` query parameter. 

This is an outstanding issue. It would seem that `catalogue` gets set on a leaf node click but it never gets read in.

> This is not removed when I click 'Return to Collections'. It also doesn't designate a catalog (it's a collection) and is not enough information to know which items it should be showing when reloaded. It sounds like you need a separate catalog path which determines what is shown in the catalogs/collections tree, and collection path which determines that we're showing an items list from a certain collection.

I think thats the idea, just a bug at the moment.

> Note that at present the catalogue path never changes after the page is loaded. This may change, but only if a user explicitly 'descends' into a sub-Catalog or chooses to go to the details page for a Collection or Item. The catalogue path determines the API endpoint the app is querying and nothing more.

I think I can see a bug with the current approach from this - if a user descends into a sub catalog and then refreshes the page, then they would never be able to get back 'out' into the root catalog as the API has requested that sub catalog as it's root.

> We might later revisit how the URL paths are generated if you don't put the catalogue path and view type in them now. The idea would be to have URLs that mirror the API structure like this:
> 
> * https://eodatahub.org.uk/catalogue/map - map view against API endpoint https://eodatahub.org.uk/api/catalogue/stac
> * (probably where 'Catalogue' will link to) https://eodatahub.org.uk/catalogue/supported-datasets/map - map view against API endpoint https://eodatahub.org.uk/api/catalogue/stac/catalogs/supported-datasets
> * https://eodatahub.org.uk/catalogue/catalogs/user-datasets/catalogs/ahayward/map - map view against my private workspace catalogue at https://eodatahub.org.uk/api/catalogue/stac/catalogs/user-datasets/catalogs/ahayward/
> * (not implemented yet) https://eodatahub.org.uk/catalogue/catalogs/supported-datasets/catalogs/ceda/collections/cmip6 - dataset details page for the CMIP6 dataset
> 
> Every entry in the catalogue then has an API URL like `https://eodatahub.org.uk/api/catalogue/stac/<catalogue path>` and a UI URL like `https://eodatahub.org.uk/catalogue/<catalogue path>` , with the two things marked `<catalogue path>` being the same. Then, in the UI, if the catalogue path identifies a STAC Catalog then appending `/map` or `/list` to it gives a map or a list of its contents.

I like this idea to try and have a bit of standardisation between the STAC FastAPI and the Resource Catalogue. Then the rest of the query params e.g. center point, zoom level, aoi can be left as query params.


---

**Issues I've noticed**
- Clearing an AOI does not clear it from the filter

File: src/context/AppContext/index.tsx [hunk 1]
```diff
@@ -22,6 +30,10 @@ const AppContext = createContext<AppContextType | undefined>(undefined);
 
 const AppProvider: React.FC<AppProviderProps> = ({ children }) => {
   const [state, dispatch] = useReducer(reducer, initialState);
+  const [urlRead, setUrlRead] = useState(false);
+
+  const { actions } = useFilters();
```
Comment (james-hinton) [inline src/context/AppContext/index.tsx:35]: `const { actions : { setTemporalFilter } } = useMap()`

File: src/context/AppContext/index.tsx [hunk 2]
```diff
@@ -31,6 +43,37 @@ const AppProvider: React.FC<AppProviderProps> = ({ children }) => {
...
+      for (let i = 0; i < floatArray.length; i += 2) {
+        coords.push(floatArray.slice(i, i + 2));
+      }
+      aoi = { coordinates: [coords], type: 'Polygon' };
+      const transformedCoords = coords.map((coord) => fromLonLat(coord));
+      const polygon = new Polygon([transformedCoords]);
+      const feature = new Feature({ geometry: polygon });
+      drawingSource.addFeature(feature);
+    }
+    actions.setTemporalFilter({
```
Comment (james-hinton) [inline src/context/AppContext/index.tsx:68]: `setTemporalFilter`

File: src/context/AppContext/index.tsx [hunk 3]
```diff
@@ -31,6 +43,37 @@ const AppProvider: React.FC<AppProviderProps> = ({ children }) => {
...
+      temporal: { start: startDate, end: endDate },
+      textQuery,
+      aoi: aoi ? aoi : null,
+      qualityAssurance: qa,
+    });
+    setUrlRead(true);
+  }, [drawingSource, actions, urlRead]);
```
Comment (james-hinton) [inline src/context/AppContext/index.tsx:75]: `}, [drawingSource, setTemporalFilter, urlRead]);`

File: src/context/MapContext.tsx [hunk 4]
```diff
@@ -84,6 +98,60 @@ export const MapProvider = ({ initialState = {}, children }: MapProviderProps) =
...
+    });
+  }, [map]);
+
+  useEffect(() => {
+    if (!map) return;
+    if (initialMapValues) return;
+    const mapView = map.getView();
+    const centre = getQueryParam('centre');
+    if (centre) {
+      mapView.setCenter([parseFloat(centre.split(',')[0]), parseFloat(centre.split(',')[1])]);
```
Comment (james-hinton) [inline src/context/MapContext.tsx:146]: For future, we should have a dedicated place for map utilities like this as theres a lot of map logic spread about

**Comment by ahaywardtvuk** [comment]
The first two bugs still happen for me. The second doesn't always happen but I've seen it happen. I can understand why the first happens but the second seems a bit odd - as if there's some cross-user caching going on.

As stated in the story, using the referer header to determine where to send the user back to after login will not work reliably. Browser strip anchors and are moving towards supplying less information in this header in more and more circumstances.

**Comment by ahaywardtvuk** [comment]
The second bug may be because `https://dev.eodatahub.org.uk/sign_in/` returns a redirect with the following headers included

```
cache-control: max-age=300
date: Thu, 14 Nov 2024 11:24:23 GMT
expires: Thu, 14 Nov 2024 11:29:23 GMT
location: https://dev.eodatahub.org.uk/oauth2/start?rd=https%3A//dev.eodatahub.org.uk/news/
referrer-policy: same-origin
vary: Cookie
```

This means the redirect can be cached _without_ the contents of the referer header being in the cache key.

This will go away if we stop using the referer header and start using a query parameter. Otherwise we'd need to change the Vary header to `vary: Cookie, Referer`.

**Comment by james-hinton** [comment]
Request for review from @IsaacJenkinsTPZ

**Pull request merged**